### PR TITLE
Add GIA candidate feedback download link to Candidate page

### PR DESCRIPTION
**Pull request by anthonydillon**
## Done

- Add functionality to allow a candidate to download their assessment documents from the candidate page

## QA

- Open the demo
- Go to /careers/application/{token} - [tokens in different states](https://pastebin.canonical.com/p/xSw9w6hhmD/)
- Check that in the Assessment stage there is a request button when appropriate
- Click the button and try to enter a random email and check the validation works
- Use the candidate's email address and submit the form
- Check that this is successful and provides one or more reports to open

## Issue / Card

Fixes https://warthogs.atlassian.net/browse/WD-3015

### CODE REVIEW

**Comment by vikorama** [comment]
The button 'request assessment report' resizes into two-three lines very quickly on mobile

image

**Comment by vikorama** [comment]
Also on mobile the modal window is positioned at the bottom of the screen (the form withdraw your application is in the center):

image

**Comment by vikorama** [comment]
I also couldn't find the button for PPA test (for the multiple assessment case I only found one Gia report), but maybe I'm missing something. In my case I got it at the end of the early stage before a talk with the HR team member, so I'd expect to find a button somewhere here..?

Frame 5

**Review by albertkol — APPROVED** [review]
Looking good overall. And QA passed too. Just some comments.

The stuff I saw in the repo is a bit confusing though, imo. Lots of helper functions in the same file where routes are defined in the same place where views function reside.

**Comment by marinacastejon** [comment]
Thanks @anthonydillon , a few comments:
- The copy is incorrect in multiple places (button, popup text, error message, etc.), please check the [copy doc](https://docs.google.com/document/d/1RZXj-LH_7Eg-pWJR9tE_V3D5CSDzCzyhyn0WoHLZwpQ/edit#bookmark=id.b3u9amy8us2c)  which contains the latest copy approved by stakeholders
- I could not check the case of a candidate with multiple reports, could you please share another candidate page with this use case? Thanks

File: webapp/application.py
Comment (anthonydillon) [reply webapp/application.py:240]: Good point on the `attachment` null check. I have left the function but restructured it to match your suggestion of the constant.

File: webapp/application.py
Comment (anthonydillon) [reply webapp/application.py:349]: Good spot :+1:

**Comment by vikorama** [comment]
> > I also couldn't find the button for PPA test (for the multiple assessment case I only found one Gia report), but maybe I'm missing something. In my case I got it at the end of the early stage before a talk with the HR team member, so I'd expect to find a button somewhere here..?
> > Frame 5
> 
> That was the initial plan but there is no way for me to identify if the candidate feedback is GIA or PPA as they have the same name. So I provide a button for all of them in the GIA stage.

 Got it! If it's not clear for users, we may want to fix it later (in a separate ticket) with a different copy.
 
 I couldn't get the PPA report still. The multiple case scenario only gives me the GIA report (see below). Should there be another green button ("Download your Thomas International GIA candidate report" and "Download your Thomas International PPA candidate report")? Or it gets downloaded in one go with one button?

https://github.com/canonical/canonical.com/assets/129760288/21c018a1-d2b7-4fef-80fc-23f4b43ee2e2

**Comment by anthonydillon** [comment]
> The copy is incorrect in multiple places (button, popup text, error message, etc.), please check the [copy doc](https://docs.google.com/document/d/1RZXj-LH_7Eg-pWJR9tE_V3D5CSDzCzyhyn0WoHLZwpQ/edit#bookmark=id.b3u9amy8us2c) which contains the latest copy approved by stakeholders

I have reviewed the mismatch in content which is due to the difficultly in identifying which attachment is which.

**Comment by anthonydillon** [comment]
> I could not check the case of a candidate with multiple reports, could you please share another candidate page with this use case? Thanks

There is a link in the QA steps of the description which has links to candidates in different states.

**Comment by anthonydillon** [comment]
> > > I also couldn't find the button for PPA test (for the multiple assessment case I only found one Gia report), but maybe I'm missing something. In my case I got it at the end of the early stage before a talk with the HR team member, so I'd expect to find a button somewhere here..?
> > > Frame 5
> > 
> > 
> > That was the initial plan but there is no way for me to identify if the candidate feedback is GIA or PPA as they have the same name. So I provide a button for all of them in the GIA stage.
> 
> Got it! (If it's not clear for users, we may want to fix it later (in a separate ticket) with a different copy.)
> 
> I couldn't get the PPA report still. The multiple case scenario only gives me the GIA report (see below). Should there be another green button ("Download your Thomas International GIA candidate report" and "Download your Thomas International PPA candidate report")? Or it gets downloaded in one go with one button?
> 
>  Screen.Recording.2023-05-15.at.14.06.23.mov

You should see multiple buttons stacked as I don't know which is which:
image

**Comment by marinacastejon** [comment]
@anthonydillon Thanks, I did check those links, and the first one that says multiple reports I still only see one button: 
image

**Comment by marinacastejon** [comment]
@anthonydillon 
I saw you updated the rest of the copy, thanks. So for the button, I will update in the copy doc the button to a generic label since it's not technically feasible to differentiate one report from the other. 
Please let me know when the case of multiple reports is ready to be checked (as I explained the pastebin examples are missing that case, I have double-checked several times). 
Also as a matter of process, if something is not feasible, the designer should be told about it, finding out during QA is not what we usually do. I will update the wireframes based on what we know is feasible. Please keep that in mind in the future, thanks.

**Comment by anthonydillon** [comment]
Sorry, I have updated the pastebin link in the description for someone with multiple documents.

**Comment by anthonydillon** [comment]
Updated the download buttons to links with the icon: 
image

**Comment by anthonydillon** [comment]
Thanks for the feedback @marinacastejon I understand its not ideal. I will block this until you can design a solution to the issue of the documents.

**Comment by marinacastejon** [comment]
Thanks for the updates. 
I what case are there 3 buttons/links? (referring to your last screenshot)

I was wondering in the case when there are multiple reports if any of these solutions are feasible: 
- Do we always get the GIA first and the PPA second? If this is the case, we could add "GIA" in the label of the first button and link, and "PPA" in the second one. It might be a hacky solution, but it would still be better than showing 2 labels exactly the same. 
- Another option is if we can detect that there are 2 reports ready instead of one, we change the button and link labels to say "Reports" in plural, and the candidate downloads both. This is worse in my opinion, but maybe it gives you another idea of other temporary solutions?
- A third option is updating the text in the Talent interview section to say they can download the report in the section "Assessment". @vikorama what do you think? This is the [miro board with the flow](https://miro.com/app/board/uXjVO9E_QFA=/?moveToWidget=3458764551533648113&cot=14)
The reason why I think will be very confusing and maybe so much so that people will still request the PPA by email to hiring leads is that if candidates have already downloaded the GIA and the page tells them they are in late stage, seeing a second button in the GIA area seems totally random and not related to the stage one is in and with the assessment one just took.

**Comment by anthonydillon** [comment]
The order of the attachments is not ordered in any way. But I do get the created date so there is some way to guess but this seems loose to me.

There are three in this case, because this candidate has two GIA results attached to their allocation. 

I tried for some time to make multiple reports download at the same time which I could not find a solution for. Seems browsers do not like multiple file downloads.

**Comment by vikorama** [comment]
@marinacastejon I'd prefer the 3rd or the 4th version. I like the idea of having a separate section, but it'd require more UX and development work.

I'd go for the 3rd one, as it seems to be easier to implement, and see how it works. Maybe we can add a line like this with (or without) an anchor link to the assessment section? 

Frame 620 (1)

**Comment by marinacastejon** [comment]
Thanks @vikorama, I agree. 
@anthonydillon I think we can unblock this but tweaking it with the third option I proposed, see Victoria's screenshot. This is, updating the text in the Talent interview section to say they can download the report in the section "Assessment".
What do you think?

**Comment by andesol** [comment]
@anthonydillon Unrelated to this PR, but you may want to know that there's something wrong in the css in production, see https://github.com/canonical/canonical.com/issues/862

**Comment by anthonydillon** [comment]
> @anthonydillon Unrelated to this PR, but you may want to know that there's something wrong in the css in production, see #862

Thanks, that is due to a Vanilla dep update. That I can try and tackle as a separate issue.

**Comment by anthonydillon** [comment]
> Thanks @vikorama, I agree. @anthonydillon I think we can unblock this but tweaking it with the third option I proposed, see Victoria's screenshot. This is, updating the text in the Talent interview section to say they can download the report in the section "Assessment". What do you think?

@marinacastejon @vikorama I have added the text to the technical assessment stage.

**Comment by vikorama** [comment]
@anthonydillon I keep getting this message for some reason:

image

**Comment by anthonydillon** [comment]
Sorry demos are down again.

**Comment by marinacastejon** [comment]
Hi @anthonydillon, thanks for implementing the fixes we discussed. 
The anchor link is not working well for me (I used the first URL in the paste bin). I tested it with another candidate from the UX roles and I can reproduce the issue. Basically, when I land on the page, I click on the anchor link and it works. If I then scroll on the page up or down, and then click again, the anchor link doesn't work. Let me know when you fix it and I'll QA it again. Thanks

**Comment by anthonydillon** [comment]
> Hi @anthonydillon, thanks for implementing the fixes we discussed. The anchor link is not working well for me (I used the first URL in the paste bin). I tested it with another candidate from the UX roles and I can reproduce the issue. Basically, when I land on the page, I click on the anchor link and it works. If I then scroll on the page up or down, and then click again, the anchor link doesn't work. Let me know when you fix it and I'll QA it again. Thanks

Sorry, you mean that when you scroll down and back up the link to the report down not work? Do you mean the button to open the modal or the download links?

**Comment by marinacastejon** [comment]
> > Hi @anthonydillon, thanks for implementing the fixes we discussed. The anchor link is not working well for me (I used the first URL in the paste bin). I tested it with another candidate from the UX roles and I can reproduce the issue. Basically, when I land on the page, I click on the anchor link and it works. If I then scroll on the page up or down, and then click again, the anchor link doesn't work. Let me know when you fix it and I'll QA it again. Thanks
> 
> Sorry, you mean that when you scroll down and back up the link to the report down not work? Do you mean the button to open the modal or the download links?

Sorry, I mean the anchor link to the Assessment section
image

**Comment by anthonydillon** [comment]
:+1: from Hanna too

**Pull request merged**
